api/global/internal/meter.go

```diff
@@ -203,6 +237,121 @@ func (bound *instHandle) Unbind() {
 	(*implPtr).Unbind()
 }
 
+// Any Observer delegation
+
+func (obs *obsImpl) setDelegate(d metric.Meter) {
+	if obs.nkind == core.Int64NumberKind {
+		obs.setInt64Delegate(d)
+	} else {
+		obs.setFloat64Delegate(d)
+	}
+}
+
+func (obs *obsImpl) unregister() {
+	unreg := obs.getUnregister()
+	if unreg != nil {
+		unreg.Unregister()
+		return
+	}
+	obs.meter.lock.Lock()
+	defer obs.meter.lock.Unlock()
+	delete(obs.meter.observers, obs)
+	if len(obs.meter.observers) == 0 {
+		obs.meter.observers = nil
+	}
+}
+
+func (obs *obsImpl) setCallback(callback interface{}) {
+	obs.lock.Lock()
+	defer obs.lock.Unlock()
+	obs.callback = callback
+}
+
+func (obs *obsImpl) getUnregister() observerUnregister {
+	ptr := atomic.LoadPointer(&obs.delegate)
+	if ptr == nil {
+		return nil
+	}
+	if obs.nkind == core.Int64NumberKind {
+		return *(*metric.Int64Observer)(ptr)
+	}
+	return *(*metric.Float64Observer)(ptr)
+}
+
+// Int64Observer delegation
+
+func (obs *obsImpl) setInt64Delegate(d metric.Meter) {
+	obsPtr := new(metric.Int64Observer)
+	obs.lock.Lock()
+	defer obs.lock.Unlock()
+	cb := obs.callback.(metric.Int64ObserverCallback)
+	*obsPtr = d.RegisterInt64Observer(obs.name, cb, obs.opts...)
+	atomic.StorePointer(&obs.delegate, unsafe.Pointer(obsPtr))
+}
+
+func (obs int64ObsImpl) SetCallback(callback metric.Int64ObserverCallback) {
+	if obs.trySetCallbackInDelegate(callback) {
+		return
+	}
```

Comment:
Are two attempts necessary? 
Wouldn't  this work?
```
obs.observer.setCallback(callback)
obs.trySetCallbackInDelegate(callback)
```

Reply:
That would work too, but it is an (unprofiled ;) ) optimization - `obs.observer.setCallback(callback)` involves locking the mutex. And I expect the first call to `trySetCallbackInDelegate` to usually succeed anyway.